### ISSUE TITLE
AssertionError: If you see this error, please let us know by opening an issue at...

### ISSUE THREAD

**Issue by jotsif** [google/jax#1919]
Hi there!

Just ran into 
```
AssertionError: If you see this error, please let us know by opening an issue at
https://github.com/google/jax/issues 
since we thought this was unreachable!
```
so I thought I better do as i was told!

I am trying to implement an ML algorithm, and when jit tracing a function I get the error above.

The issue was that i used a good old numpy randint inside the loss function, and when replacing that with a jax.random randint the error went away.

**Comment by mattjj** [comment]
Thanks for raising this! However, it's not very actionable for us without a bit more information. Can you share a repro (smaller is better)?

**Comment by jotsif** [comment]
Yes! A branch where the error is left is here: https://github.com/jotsif/fm_jax/tree/jit_error, and run `warp_jit_error.py` to reproduce.

**Comment by mattjj** [comment]
Thanks for the kind words, and the repro! I'll look into it.

By the way, as a workaround if you want to dodge this error, you can put this at the top of your file:

```python
from jax.interpreters import partial_eval as pe
pe._thread_local_state.remat = True
```

[The error](https://github.com/google/jax/blob/82dbf9131105a0b3a22c191930bc42b15d420794/jax/interpreters/partial_eval.py#L598-L608) is a "voluntary" and [temporary](https://github.com/google/jax/blob/82dbf9131105a0b3a22c191930bc42b15d420794/jax/interpreters/partial_eval.py#L512) one that I inserted to check our understanding that a change we made in #1749 wouldn't cause any extra (but otherwise harmless) FLOPs to be performed. The above lines just disable the "voluntary" error-checking mechanism.

I want to look into this example to understand if it's an edge case (which we can just patch) or if I should revert the extra-FLOP-potential part of #1749 that this error is meant to check.

**Comment by mattjj** [comment]
I finally got to looking at this, and I'm glad you filed it: our control flow primitives had cases where they were letting concrete abstract values through their abstract eval rules. That's not incorrect behavior, but it is different from what we expected when we put in this error check for extra FLOPs.

The fix should be in #1958 !
